src/subprocess-posix.cc

```diff
@@ -169,11 +171,34 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+void SubprocessSet::HandleChildDeath() {
+  if (interrupted_ != SIGCHLD)
+    return;
+  interrupted_ = false;
+  int options = WEXITED | WNOWAIT | WNOHANG;
+  for (vector<Subprocess*>::iterator i = running_.begin();
+       i != running_.end(); ) {
+    siginfo_t info = {0};
+    int res = waitid(P_PID, (id_t)(*i)->pid_, &info, options);
```

Comment:
Minor: I think that if waitid(-1, ..) would be used here instead, you could loop over all finished processes in the outer loop and iterating running_ in the inner, making number of waitid() syscalls ~finished processes, which would be marginally more efficient.